### PR TITLE
fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch

### PR DESCRIPTION
**Pull request by bdhirsh**
Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85844
* #85875
* #85681
* #82601
* #82602
* __->__ #82771
* #83265

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
okey dokey

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -943,6 +943,12 @@
   device_check: NoCheck   # TensorIterator
   variants: function, method
   tags: nondeterministic_seeded
+  dispatch:
+    # This op needs to be a primitive w.r.t. Python,
+    # because functionalization wants to convert bernoulli_.float -> bernoulli.p.
+    # If we leave it as CompositeImplicitAutograd, then we'll run this kernel
+    # If we leave it as CompositeImplicitAutograd, then we'll run this kernel
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:950]: duplicated line in comment?

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -333,6 +333,10 @@
   p: zeros_like(p)
   result: self_t.zero_()
 
+- name: bernoulli.p(Tensor self, float p, *, Generator? generator=None) -> Tensor
+  self: zeros_like(grad)
+  result: auto_element_wise
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml:338]: The behavior today is that bernoulli.p returns a Tensor without requires_grad. Does this PR change the behavior by giving it a backwards formula?
```
In [6]: x = torch.randn(3, 3, requires_grad=True)

In [7]: y = torch.bernoulli(x, 0.5)

In [8]: y
Out[8]:
tensor([[1., 1., 0.],
        [0., 1., 0.],
        [1., 0., 0.]])

```

Comment (albanD) [reply tools/autograd/derivatives.yaml:338]: Correct, please mark `output_differentiability = [False]` here to preserve that behavior

Comment (bdhirsh) [reply tools/autograd/derivatives.yaml:338]: whoops, thanks :)

**Review by albanD — APPROVED** [review]
Duplicate comments needs to be removed. But good otherwise.

**Pull request closed without merging**